### PR TITLE
Check address validity of transactions

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR adds two spam prevention measures for transactions:
* Check address validity for all transactions (essentially check the checksum)
* Check that the address index is not obviously non-existent by checking its index. Later this will track `AccountCreated` events to keep track of the latest created account, and add a buffer. For now we use a fixed maximum index.

### CODE REVIEW

File: utils/unittest/fixtures.go [hunk 1]
```diff
@@ -37,6 +37,15 @@ func RandomAddressFixture() flow.Address {
 	return addr
 }
 
+func InvalidAddressFixture() flow.Address {
```
Comment (tarakby) [inline utils/unittest/fixtures.go:40]: 👌

File: engine/collection/ingest/config.go [hunk 2]
```diff
@@ -27,6 +29,7 @@ func DefaultConfig() Config {
 		MaxGasLimit:           flow.DefaultMaxGasLimit,
 		AllowUnknownReference: false,
 		CheckScriptsParse:     true,
+		MaxAddressIndex:       1_000_000,
```
Comment (tarakby) [inline engine/collection/ingest/config.go:32]: Just to get an idea, do we know what's the current index in Mainnet/Testnet ?

File: access/validator.go [hunk 3]
```diff
@@ -192,6 +204,33 @@ func (v *TransactionValidator) checkCanBeParsed(tx *flow.TransactionBody) error
 	return nil
 }
 
+func (v *TransactionValidator) checkAddresses(tx *flow.TransactionBody) error {
+
+	for _, address := range append(tx.Authorizers, tx.Payer) {
```
Comment (tarakby) [inline access/validator.go:209]: Isn't the payer's responsibility to check the authorizers are valid? I mean shouldn't the payer still pay for a transaction that has invalid authorizers?

**Review by tarakby — COMMENTED** [review]
Clean work, looks good to me 👌 
I just left a question about checking authorizers.

File: engine/collection/ingest/config.go (hunk 2 above)
Comment (jordanschalm) [reply engine/collection/ingest/config.go:32]: Posted in #flow-integration to ask. There should be ~13,000 from CoinList, and I think something like 10-20k-ish from Dapper.

Comment (jordanschalm) [reply engine/collection/ingest/config.go:32]: Checked and 2 hours ago we had 27272 accounts.

File: access/validator.go (hunk 3 above)
Comment (jordanschalm) [reply access/validator.go:209]: It is the payer's responsibility to some extent, but we still also want to not include invalid transactions as well. There is a whole class of spam where payers sign an invalid transaction for which they don't have enough funds to pay. Plus, be rejecting these transactions here we reduce overall load on the rest of the network. In general the collection nodes should filter out as much invalid inputs as they can.

Comment (tarakby) [reply access/validator.go:209]: For payers with enough balance, it makes sense to make them pay as a countermeasure of spamming collectors with invalid transactions. I agree we should filter the payers without balance early in the pipeline.
Till we implement collector's access to the execution state, I think your approach is right as I guess a spamming payer is more likely to be an account with low/no balance.

**Review by tarakby — APPROVED** [review]

File: access/validator.go (hunk 3 above)
Comment (jordanschalm) [reply access/validator.go:209]: > For payers with enough balance, it makes sense to make them pay as a countermeasure of spamming collectors with invalid transactions

Good point 👍

**Review by ramtinms — APPROVED** [review]
Looks good to me

**Pull request merged**